### PR TITLE
move the detect n+1 hydration to dev ns

### PR DESCRIPTION
**Pull request by qnkhuat**
added in #40272, but I figure it's better if we leave this code in the dev ns

### CODE REVIEW

**Review by sansolovyov — APPROVED** [review]
Makes sense

File: dev/src/dev.clj [hunk 1]
```diff
@@ -297,6 +298,27 @@
...
+    (t2/with-call-count [call-count]
+      (let [res (next-method model strategy k instances)
+            ;; if it's a lazy-seq then we need to realize it so call-count is counted
+            res (if (instance? clojure.lang.LazySeq res)
+                  (doall res)
+                  res)]
+        ;; only throws an exception if the simple hydration makes a DB call
+        (when (pos-int? (call-count))
+          (throw (ex-info (format "N+1 hydration detected!!! Model %s, key %s]" (pr-str model) k)
+                          {:model model :strategy strategy :k k :items-count (count instances) :db-calls (call-count)})))
```
Comment (crisptrutski) [inline dev/src/dev.clj:319]: nitpick: I'd break this map up into multiple lines

File: dev/src/dev.clj [hunk 2]
```diff
@@ -297,6 +298,27 @@
                                  (qp.compile/compile built-query))]
     (into [query] params)))
 
+(methodical/defmethod t2.hydrate/hydrate-with-strategy :around ::t2.hydrate/multimethod-simple
+  "Throws an error if do simple hydrations that make DB call on a sequence."
+  [model strategy k instances]
+  (if (or config/is-prod?
```
Comment (crisptrutski) [inline dev/src/dev.clj:304]: Do we still need this check if we register this method this in dev?

**Comment by crisptrutski** [comment]
I'm a bit torn about this one, as I often develop without bothering with the `dev` namespace.

Maybe I need to just get with the program :-)

**Review by crisptrutski — APPROVED** [review]

File: dev/src/dev.clj (hunk 2 above)
Comment (piranha) [reply dev/src/dev.clj:304]: Better safe than sorry :P

**Comment by piranha** [comment]
> I'm a bit torn about this one, as I often develop without bothering with the dev namespace.

Well it's easy: just don't write N+1 code, that's it. :)

File: dev/src/dev.clj (hunk 1 above)
Comment (qnkhuat) [reply dev/src/dev.clj:319]: oops sr I auto merged this already.. will fix it in one of my other PRs

File: dev/src/dev.clj (hunk 2 above)
Comment (qnkhuat) [reply dev/src/dev.clj:304]: if you start your repl with `:dev` alias, then you won't miss anything.

**Comment by qnkhuat** [comment]
@crisptrutski not sure how you start your reply but I'll always have `:dev` alias, and the first thing I do is to eval `(dev)` which set me into `dev` namespace and load all the essentials

**Comment by crisptrutski** [comment]
> > I'm a bit torn about this one, as I often develop without bothering with the dev namespace.
> 
> 
> 
> Well it's easy: just don't write N+1 code, that's it. :)

Don't we want it to get caught in tests though?

**Pull request merged**